### ISSUE TITLE
Auto column width and row height with multi-line cell values

### ISSUE THREAD

**Issue by nachtkinder** [DeepSpace2/StyleFrame#87]
Hi,

I'm having an issue with the auto determined column width and row height when using StyleFrame with a pandas DataFrame that contains multi-line values.

```python
import pandas as pd
from styleframe import StyleFrame

df = pd.DataFrame({"a": ("Line 1\nLine 2\nLine 3",)})
sf = StyleFrame(df)
sf.to_excel("test.xlsx", best_fit=list(df.columns), index=False).save()
```
What I expect test.xlsx to look like is:
```
   a   |  b
-------|------
Line 1 |Line 1
Line 2 |Line 2
Line 3 |Line 3
-------|------
```

What it actually looks like is:
```
         a         |         b
-------------------|------------------
      Line 2       |      Line 2
-------------------|------------------
```

"Line 1" and "Line 3" are present in each cell but the row is not rendered with sufficient height for them to be visible.

This behaviour is the same regardless of any wrap_text or width properties set on the columns.

The README.md states:
> The StyleFrame object will auto-adjust the columns width and the rows height but they can be changed manually

The API documentation for StyleFrame.to_excel states the formula used to determine a column's width as:
```python
(len(longest_value_in_column) + A_FACTOR) * P_FACTOR
```

...which results in column widths a little bigger than the largest combined length of all lines plus the newline characters between them in a single value.

Is there a way to instruct StyleFrame to treat multi-line value differently such that:

a. Columns that contain multi-line values are not much wider than the longest *line* in a value contained within them

and

b. Rows that contain multi-line values have sufficient height to display all lines of any cell contained within

Versions
---------

Python 3.7.5
StyleFrame 3.0.5
pandas 1.0.4
openpyxl 2.6.3

Described behaviour also present with pandas 1.1.3 and openpyxl 3.0.5

**Comment by DeepSpace2** [comment]
Hi.

> a. Columns that contain multi-line values are not much wider than the longest line in a value contained within them

This is quite easily fixable, see the commit in which I referenced this issue. This will each row in each column on `\n` and use the length of the longest line in each row as the length for the calculation.

> b. Rows that contain multi-line values have sufficient height to display all lines of any cell contained within

This is a bit tricker because I suspect this is going to be more font-size sensitive.

For now, you can use the following workaround before calling `.to_excel`:

```python
df = pd.DataFrame({"a": ("Line 1\nLine 2\nLine 3",)})
sf = StyleFrame(df)
sf.set_row_height_dict({index + 2: (max(sf.data_df[column].astype(str).str.count('\n') for column in sf.data_df.columns) + sf.A_FACTOR) * sf.P_FACTOR * 2
                        for index, _ in enumerate(sf.data_df.index)})
sf.to_excel("test.xlsx", best_fit=list(df.columns), index=False).save()
```

**Comment by fiendish** [comment]
I don't think that set_row_height_dict works. I had to do
```Python
sf.set_row_height_dict(dict(
        zip(sf.data_df.index.astype(int)+2, (pandas.concat([sf.data_df[column].astype(str).str.count('\n') for column in sf.data_df.columns], axis=1).max(axis=1) + sf.A_FACTOR * sf.P_FACTOR * 2).values)
    ))
```

**Comment by fiendish** [comment]
I think rather you would want to multiply the number of lines (count("\n")+1) by A_FACTOR and then add 2x P_FACTOR afterward.

**Comment by DeepSpace2** [comment]
@fiendish

>  multiply the number of lines (count("\n")+1) by a padded character height

That's essentially what `A_FACTOR` and `P_FACTOR` are:

> `best_fit` will attempt to calculate the correct column-width based on the longest value in each provided column. However this isn’t guaranteed to work for all fonts (works best with monospaced fonts). The formula used to calculate a column’s width is equivalent to `(len(longest_value_in_column) + A_FACTOR) * P_FACTOR`.  The default values for `A_FACTOR` and `P_FACTOR` are `13` and `1.3` respectively, and can be modified before calling `StyleFrame.to_excel` by directly modifying `StyleFrame.A_FACTOR` and `StyleFrame.P_FACTOR`.

*from [the docs](https://styleframe.readthedocs.io/en/latest/styleframe.html?highlight=A_FACTOR#styleframe.StyleFrame.to_excel)*

The above suggested solution does what it says on the tin :)

This

```python
df = pd.DataFrame({"a": ("Line 1\nLine 2\nLine 3",)})
sf = StyleFrame(df)
sf.to_excel("test.xlsx", best_fit=list(df.columns), index=False).save()
```

generates

![image](https://user-images.githubusercontent.com/6841988/123341611-ee7ceb80-d556-11eb-817b-3b8e889526fc.png)

however

```python
df = pd.DataFrame({"a": ("Line 1\nLine 2\nLine 3",)})
sf = StyleFrame(df)
sf.set_row_height_dict({index + 2: (max(sf.data_df[column].astype(str).str.count('\n') for column in sf.data_df.columns) + sf.A_FACTOR) * sf.P_FACTOR * 2
                        for index, _ in enumerate(sf.data_df.index)})
sf.to_excel("test.xlsx", best_fit=list(df.columns), index=False).save()
```

generates

![image](https://user-images.githubusercontent.com/6841988/123341667-081e3300-d557-11eb-959a-a6d23541ea8c.png)

**Comment by coolbeam** [comment]
Hi
I got an error when processing my data. I guess when the table has multiple rows, this code may not work. 
`ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().`
Maybe it is because the max operation in `sf.set_row_height_dict()` line. But I don't know.

Another problem is in `apply_style_by_indexes`, when no grid or index satisfy the provided condition, there is a key error.
`KeyError: "None of [Index([.......dtype='object')] are in the [columns]"`

**Comment by DeepSpace2** [comment]
@coolbeam Indeed, `sf.data_df[column].astype(str).str.count('\n')` should be `sf.data_df[column].astype(str).str.count('\n').max()`.

As for your second point, I think that should be a separate issue as I'm not sure this is related to this. Anyway, please provide a reproducible example as I'm not sure I follow.

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
Maybe related #131
